### ISSUE TITLE
Can't import Android .db file extracted from an Android STB

### ISSUE THREAD

**Issue by AE-Ismail** [PredatH0r/ChanSort#332]
Is it possible to add support for .db files extracted from Android STB?
I attached the file for anyone to try to open it.

[database.zip](https://github.com/PredatH0r/ChanSort/files/10273569/database.zip)

**Comment by PredatH0r** [comment]
What's the brand and model name of your STB?
The .db file is a Sqlite database file and can be viewed/edited with tools like SqliteBrowser or Sqlite Studio.
ChanSort already supports several other channel list formats based on Sqlite, but they have a different internal structure.
So technically it's possible to add support for it, but I'm on a very tight schedule at the moment and can't tell if/when I find the time for it.

**Comment by AE-Ismail** [comment]
My device is Vision Edge 4K, running Android 7.1 with hi3798mv200 Chipset.

**Comment by AE-Ismail** [comment]
>The .db file is a Sqlite database file and can be viewed/edited with tools like SqliteBrowser or Sqlite Studio.

I tried Sqlite Studio, and it actually opened the file, but I couldn't export the file back to .db.
And also I couldn't rearrange the channels as in ChanSort.

**Comment by PredatH0r** [comment]
My apologies...
These tools aren't specialized channel editors, but merely tools for developers to look at the data stored inside a Sqlite database file and provide some means to modify them. In "data view" similar to Excel or in a "sql query view" with specific database commands.
They are not really designed for end-users.

I hope I find some time over the holidays to add support for your list.

**Comment by AE-Ismail** [comment]
Yes, that's what I noticed with that tool, it was too complicated for me, an end-user.
Thank you for your effort, really appreciate it.

**Comment by AE-Ismail** [comment]
Hello again,

Just a kind reminder to see if you'll have time to support Android STB Channel List.
This screenshot is from your latest release, 10/01/2023.
Thank you.

![2023-01-10_20-59-42](https://user-images.githubusercontent.com/118636398/211638408-4f6be872-7c41-4a9e-baaa-e5774bbef43a.png)

**Comment by PredatH0r** [comment]
Not yet, but it's up on my ToDo list. I hope I find the time over the weekend.

**Comment by AE-Ismail** [comment]
Okay. Thank you again, and have a great weekend.

**Comment by PredatH0r** [comment]
There is now a version on https://chansort.com/ChanSort_2023-01-14.zip which can edit your file.

I had to disable editing favorites because I need more information on how it works.
Also, I can only support DVB-S at the moment. Based on the file content, it seems that the box also supports DVB-T and DVB-C.

Do you have a way to scan for DVB-T or DVB-C channels too? A file containing DVB-S plus DVB-C or -T channels would give me the information I need to support those.
Could you add some channels to favorite list 1-4 on your TV's menu, export the file and send it to me? I can then also add support for editing favorites.

Thanks in advance,
Horst

**Comment by PredatH0r** [comment]
Also, could you please add a few TV and a few radio channels to the favorite list?
That's another open question whether TV and radio are treated separately (both fav lists start at 1) or if it's a combine list.

**Comment by AE-Ismail** [comment]
Thank you so much for your help and time in adding support to this file.
As for your first question about DVB-T or DVB-C, the box doesn't have the tuner in its hardware and has the DVB-S tuner only.

As for the favorites lists, I added some channels to favorites on different satellites on TV and Radio, hope it helps you out.

Another point I noticed was that I have 8 different satellites, and the program shows a list of all the channels at once as if they were on 1 satellite, although it sorts them according to satellite, but I couldn't find a way to view one satellite at a time. Is this normal with multiple satellite files, or is it specific to my file type?

Also, I attached a copy of the .db file after editing the favorites list.

Much obliged,
Ahmed Ismail

[database.zip](https://github.com/PredatH0r/ChanSort/files/10418040/database.zip)


![2023-01-14_17-29-26](https://user-images.githubusercontent.com/118636398/212479775-39fdc151-e2bb-4e8b-9ddc-e7975da06574.png)

**Comment by PredatH0r** [comment]
Thanks for the file!
I could now find out that some of my assumptions were wrong.
Does your TV allow to add 1 channel to multiple favorite lists? e.g. to "News" and "Education"?
The database format would allow it, but the data I see in your file now suggests that the receiver's software might not support it.

In the database all channels are in a single list with continuous program numbers across all satellites.
Your first file however ordered in a way where all TV channels came first, grouped by Hotbird 13, Nilesat, Amos 3/7, Eutelsat 9B and then Radio/data channels also in a group per satellite.
The database would allow mixing TV and radio channels from all satellites in any order. But the firmware of the box might have certain restrictions/expectations on the data, that I can't tell. Like keeping the channels grouped together by TV/Radio and sat.

How does that work on the TV? Do you have to select a specific satellite first before you can enter it's number on the remote control? Or is the number enough and it jumps to the right satellite automatically? 
Does it start with program number 1 for each satellite list or does each satellite start at whatever the last number was in the previous satellite ?

The 2nd file you just sent me, was that already reordered with ChanSort, mixing the various satellites, or was that only edited on the TV?

Sorry for so many questions, but it's impossible to infer that just based on the data file.
Regards,
Horst

**Comment by AE-Ismail** [comment]
Hello,

For the favorites list, the box can allow adding the same channel to multiple favorite categories:

![WhatsApp Image 2023-01-14 at 19 07 34](https://user-images.githubusercontent.com/118636398/212485883-88b52dc7-642b-43bd-8095-e4087d086a71.jpg)

As for the multiple satellites, the box groups each satellite and its channel to a sub-list after selecting that specific satellite from the remote.
![WhatsApp Image 2023-01-14 at 19 07 33vv](https://user-images.githubusercontent.com/118636398/212485941-3c1a46ad-e9e2-4878-8407-81bfd38934a2.jpg)

Then starts each satellite from number 1 for TV channels and number 1 for radio channels.

![WhatsApp Image 2023-01-14 at 19 07 33](https://user-images.githubusercontent.com/118636398/212486127-efa68e8f-d08f-4c3a-861c-181ae6084fff.jpg)
![WhatsApp Image 2023-01-14 at 19 07 33d](https://user-images.githubusercontent.com/118636398/212486134-c37f412b-0bba-443a-8d81-81fed6608d87.jpg)

The file I previously sent was extracted from the box directly after editing the favorites list, however, I didn't use it before, so I can't assume what it does to the extracted database file.

Hope that answers your questions, and please feel free to ask for any updates that could help.

Much obliged,
Ahmed

**Comment by PredatH0r** [comment]
Thanks, that's very helpful.
The way I see it is that the TV doesn't show those sequential numbers at all, that are stored in the database and that ChanSort shows as Pr#.
Instead the TV reads the channels by that order but assigns different numbers to them, starting at 1 with each satellite. And probably TV channels first, then the radio channels.
I'll prepare a new ChanSort version that keeps separate lists for each satellite then.

**Comment by AE-Ismail** [comment]
Yes, as you clarified, each satellite starts at 1 with TV channels and then radio channels.
Thank you very much for this effort, really appreciate it.

**Comment by PredatH0r** [comment]
Almost done.
Last questions: In your "Movies" favorite list, do you see channels from all satellites - or is there a separate "Movie" fav list for each satellite?

Currently I have an implementation that has a combined Fav-List, including Hotbird and Bulgariasat channels:
![image](https://user-images.githubusercontent.com/6238104/212491431-fff7b0d2-e318-47fe-9238-467ece4b09b2.png)

(The "Movies" list starts at 1 with Novasports1HD, but I scrolled down to make the screenshot smaller)

**Comment by AE-Ismail** [comment]
Thank you for this swift assistance 😊.

It seems that there's a glitch in the software of my box. Whenever I press the Fav button on the remote control, it opens the list of all the channels of the selected satellite.
But in general, all devices of the same company treat favorites of each satellite individually and doesn't combine them.

Ahmed Ismail
Subtitler & Audiovisual Translator
+201000267865
________________________________

**Comment by AE-Ismail** [comment]
Update:

I asked a friend that has a similar box with a different UI, he said that the Fav button shows all favorites across different satellites in one menu.

**Comment by PredatH0r** [comment]
New version available on https://chansort.com/ChanSort_2023-01-14.zip

For each favorite list, it allows TV and radio channels from all satellites to be added.
When saving the file, this version makes sure that TV channels come first, then all radio channels.

What I don't know is if you can arbitrarily mix channels from the various satellites (like 1 from Nilesat, 2 from Eutelsat, 3 from Nilesat again, 4 from Astra, ....) or if the box will automatically arrange all channels from the same satellite together, within each favorite list.
This version allows an arbitrary order. Maybe the box reorders them by-satellite when it imports the file.

**Comment by AE-Ismail** [comment]
Thank you so much for this update, I'll edit the list of channels and the favorite list, then import the file to the box to see if everything will be okay.

Thank you again for this addition and for your time.

**Comment by AE-Ismail** [comment]
After sorting the list, I imported the exported file and loaded it to the TV, and it worked very well with no problems at all.
The favorites list worked okay in the program, but on the TV it didn't work, I think it's a glitch in the TV software. Anyway, I didn't use it before at all, so it won't bother me that much.

There are many users with the same firmware on different devices, I'll send the program to them to check their files if it works well.

I'd like to thank you again for this great effort and help, really appreciate it. Thank you for your time.

**Comment by AE-Ismail** [comment]
Hello again, PredatH0r,

I have another .db file from another android receiver with a different processor, Amlogic S905x3 Hybrid, can it be added to the program?

Thank you.

[amdb2023_03_11_18_32.zip](https://github.com/PredatH0r/ChanSort/files/11009828/amdb2023_03_11_18_32.zip)
